common/functions/src/scalars/maths/sign.rs

```diff
@@ -51,67 +50,49 @@ impl SignFunction {
     }
 }
 
-impl Function for SignFunction {
+fn sign<S>(value: S::RefType<'_>) -> i8
+where
+    S: Scalar + Default,
+    for<'a> S::RefType<'a, ScalarType = S>: PartialOrd,
+    for<'a> S: Scalar<RefType<'a> = S>,
+{
+    match value.partial_cmp(&S::default()) {
+        Some(std::cmp::Ordering::Greater) => 1,
+        Some(std::cmp::Ordering::Less) => -1,
+        _ => 0,
+    }
+}
+
+impl Function2 for SignFunction {
     fn name(&self) -> &str {
         &*self.display_name
     }
 
-    fn return_type(&self, args: &[DataTypeAndNullable]) -> Result<DataTypeAndNullable> {
-        let nullable = args.iter().any(|arg| arg.is_nullable());
-        let data_type = if matches!(
-            args[0].data_type(),
-            DataType::UInt8
-                | DataType::UInt16
-                | DataType::UInt32
-                | DataType::UInt64
-                | DataType::Int8
-                | DataType::Int16
-                | DataType::Int32
-                | DataType::Int64
-                | DataType::Float32
-                | DataType::Float64
-                | DataType::String
-                | DataType::Null
-        ) {
-            Ok(DataType::Int8)
-        } else {
-            Err(ErrorCode::IllegalDataType(format!(
-                "Expected numeric types, but got {}",
-                args[0]
-            )))
-        }?;
-        Ok(DataTypeAndNullable::create(&data_type, nullable))
+    fn return_type(&self, args: &[&DataTypePtr]) -> Result<DataTypePtr> {
+        assert_numeric(args[0])?;
+        Ok(i8::to_data_type())
     }
 
-    fn eval(&self, columns: &DataColumnsWithField, _input_rows: usize) -> Result<DataColumn> {
-        let result = columns[0]
-            .column()
-            .to_minimal_array()?
-            .cast_with_type(&DataType::Float64)?
-            .f64()?
-            .apply_cast_numeric(|v| {
-                if v > 0_f64 {
-                    1_i8
-                } else if v < 0_f64 {
-                    -1_i8
-                } else {
-                    0_i8
-                }
-            });
-        let column: DataColumn = result.into();
-        Ok(column)
+    fn eval(&self, columns: &ColumnsWithField, _input_rows: usize) -> Result<ColumnRef> {
+        with_match_primitive_type_id!(columns[0].data_type().data_type_id(), |$S| {
```

Comment:
@PsiACE   We can use this macro here to avoid casting into Float64 in `expr` function.

Just add a bound in trait:   S: AsPrimitive<f64>

Reply:
got it

Reply:
PTAL https://github.com/datafuselabs/databend/pull/4094/commits/0ec4e3588df133090cad28ed87b7df30e934be92